File: core/GenerateParentheses/GenerateParentheses.py
from typing import List

def generateParenthesis(n: int) -> List[str]:
    res = ["()"]
    if n == 1:
        return res

    for _ in range(1, n):
        remain, temp_res = {}, []
        for item in res:
            length = len(item)
            for i in range(length + 1):
                com_str = item[0:i] + "()" + item[i:length + 1]
                if remain.get(com_str, True):
                    temp_res.append(com_str)
                    remain[com_str] = False
        res = temp_res

    return res

File: core/GenerateParentheses/GenerateParentheses.java
package com.dbc;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GenerateParentheses {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<String>(){{
            add("()");
        }};
        if (n == 1) return res;
        for (int i = 1; i < n; i++) {
            Map<String, Boolean> remain = new HashMap<>();
            List<String> tempRes = new ArrayList<>();
            for (String item : res) {
                if (remain.getOrDefault("()" + item, true)) {
                    tempRes.add("()" + item);
                    remain.put("()" + item, false);
                }
                if (remain.getOrDefault(item + "()", true)) {
                    tempRes.add(item + "()");
                    remain.put(item + "()", false);
                }
                for (int j = 1; j < item.length(); j++) {
                    String comStr = item.substring(0, j) + "()" + item.substring(j, item.length());
                    if (remain.getOrDefault(comStr, true)) {
                        tempRes.add(comStr);
                        remain.put(comStr, false);
                    }
                }
            }
            res = tempRes;
        }
        return res;
    }
}
